crates/storage/provider/src/providers/snapshot/writer.rs

```diff
@@ -87,61 +82,63 @@ impl<'a> SnapshotProviderRW<'a> {
-            // This can be a scenario where we pruned all blocks from the static file, including the
-            // genesis block.
-            if user_header.block_end() == 0 && self.writer.rows() == 0 {
-                max_block = None
+    /// Updates the `self.reader` internal index.
+    fn update_index(&self) -> ProviderResult<()> {
+        // We find the maximum block of the segment by checking this writer's last block.
+        //
+        // However if there's no block range (because there's no data), we try to calculate it by
+        // substracting 1 from the expected block start, resulting on the last block of the
+        // previous file.
+        //
+        // If that expected block start is 0, then it means that there's no actual block data, and
+        // there's no snapshotted block data.
+        let segment_max_block = match self.writer.user_header().block_range() {
+            Some(block_range) => Some(block_range.end()),
+            None => {
+                if self.writer.user_header().expected_block_start() > 0 {
+                    Some(self.writer.user_header().expected_block_start() - 1)
+                } else {
+                    None
+                }
             }
         };
 
-        max_block
+        self.reader.update_index(self.writer.user_header().segment(), segment_max_block)
     }
 
     /// Allows to increment the [`SegmentHeader`] end block. It will commit the current snapshot,
     /// and create the next one if we are past the end range.
     ///
     /// Returns the current [`BlockNumber`] as seen in the static file.
     pub fn increment_block(&mut self, segment: SnapshotSegment) -> ProviderResult<BlockNumber> {
-        debug_assert!(
-            (self.writer.rows() as u64 + self.writer.user_header().block_end()) != 0
-            || !segment.is_headers(),
-            "This function should only be called by append_header when dealing with SnapshotSegment::Headers.");
-
-        let last_block = self.writer.user_header().block_end();
-        let writer_range_end = find_fixed_range(last_block).end();
-
-        // We have finished the previous snapshot and must freeze it
-        if last_block + 1 > writer_range_end {
-            // Commits offsets and new user_header to disk
-            self.commit()?;
-
-            // Opens the new snapshot
-            let (writer, data_path) = Self::open(segment, last_block + 1, self.reader.clone())?;
-            self.writer = writer;
-            self.data_path = data_path;
-
-            let block_start = find_fixed_range(last_block + 1).start();
-            *self.writer.user_header_mut() = SegmentHeader::new(
-                SegmentRangeInclusive::new(block_start, block_start),
-                None,
-                segment,
-            )
-        } else {
-            self.writer.user_header_mut().increment_block()
+        if let Some(last_block) = self.writer.user_header().block_end() {
```

Comment:
can we use expected block range here now instead?

Reply:
yes, but not on this line, on the one below

Reply:
ye